connection.go

```diff
@@ -483,9 +527,27 @@ func (conn *OCI8Conn) ociDateTimeToTime(dateTime *C.OCIDateTime, useOCITimeZone
 		return nil, err
 	}
 
-	if !useOCITimeZone {
-		// return Go Time with conn.location
-		aTime := time.Date(int(year), time.Month(month), int(day), int(hour), int(min), int(sec), int(fsec), conn.location)
+	if !ociDateTimeHasTimeZone {
+
+		conn.timeLocationRWMutex.RLock()
+		if conn.timeLocationNextUpdate.Before(time.Now().UTC()) {
+			conn.timeLocationRWMutex.RUnlock()
+			conn.timeLocationRWMutex.Lock()
+			if conn.timeLocationNextUpdate.Before(time.Now().UTC()) {
+				conn.timeLocationNextUpdate = time.Now().UTC().Add(5 * time.Minute)
+				go func() {
+					err := conn.getDatabaseTimezone()
+					if err != nil {
+						conn.logger.Println("getDatabaseTimezone error: " + err.Error())
+					}
+				}()
```

Comment:
Why this is called from goroutine?

Reply:
So that it does not block the return of the query data

Reply:
conn.timeLocation seems not be atomic.

Reply:
It is a pointer, should either point to the last one or the new one. The database timezone should not change often. Almost never.

Reply:
Go's pointer is not atomic.

Reply:
So atomic is needed for something like foo = foo + 1. In this case we keep updating the variable with a new pointer. It does not matter if we get the new or old pointer because the timezone almost never changes.

Reply:
Assignment of pointer is not atomic. This is a reason that sync/atomic.ReadPointer is provided. So the assignment of timeLocation should be guard with mutex.

Reply:
Where does it say that assignment of a pointer is not atomic?

Here is a test program

```
package main

import (
	"fmt"
	"math/rand"
	"runtime"
	"time"
)

var aTime *time.Time

func main() {
	temp := time.Date(2009, 11, 17, 20, 34, 58, 651387237, time.UTC)
	aTime = &temp

	for i := 0; i < 50000; i++ {
		go func() {
			time.Sleep(time.Duration(rand.Intn(6000000)) * time.Nanosecond)
			temp := time.Date(2000, 10, 10, 10, 10, 10, rand.Intn(999999999), time.UTC)
			aTime = &temp
		}()
	}

	for i := 0; i < 20; i++ {
		go func() {
			time.Sleep(time.Duration(rand.Intn(6000000)) * time.Nanosecond)
			fmt.Println(runtime.NumGoroutine(), aTime)
		}()
	}

	for i := 0; i < 10; i++ {
		time.Sleep(10 * time.Millisecond)
		fmt.Println(runtime.NumGoroutine(), aTime)
	}

}
```

Output:

```
37955 2000-10-10 10:10:10.439779449 +0000 UTC
37064 2000-10-10 10:10:10.142844201 +0000 UTC
37099 2000-10-10 10:10:10.570328208 +0000 UTC
37078 2000-10-10 10:10:10.111158623 +0000 UTC
37036 2000-10-10 10:10:10.756194502 +0000 UTC
36939 2000-10-10 10:10:10.402894583 +0000 UTC
36903 2000-10-10 10:10:10.857173035 +0000 UTC
36810 2000-10-10 10:10:10.004320605 +0000 UTC
36386 2000-10-10 10:10:10.791299641 +0000 UTC
36350 2000-10-10 10:10:10.138234486 +0000 UTC
35867 2000-10-10 10:10:10.674739367 +0000 UTC
35278 2000-10-10 10:10:10.081883364 +0000 UTC
35076 2000-10-10 10:10:10.201068389 +0000 UTC
34539 2000-10-10 10:10:10.718022719 +0000 UTC
34384 2000-10-10 10:10:10.851401656 +0000 UTC
29239 2000-10-10 10:10:10.728027451 +0000 UTC
19631 2000-10-10 10:10:10.809184678 +0000 UTC
19043 2000-10-10 10:10:10.239365282 +0000 UTC
13026 2000-10-10 10:10:10.74068439 +0000 UTC
7828 2000-10-10 10:10:10.50556048 +0000 UTC
7262 2000-10-10 10:10:10.893470382 +0000 UTC
5737 2000-10-10 10:10:10.210114846 +0000 UTC
5654 2000-10-10 10:10:10.543104683 +0000 UTC
5461 2000-10-10 10:10:10.304137212 +0000 UTC
5412 2000-10-10 10:10:10.202877712 +0000 UTC
5109 2000-10-10 10:10:10.885889841 +0000 UTC
1 2000-10-10 10:10:10.370687929 +0000 UTC
1 2000-10-10 10:10:10.370687929 +0000 UTC
1 2000-10-10 10:10:10.370687929 +0000 UTC
1 2000-10-10 10:10:10.370687929 +0000 UTC
```

Reply:
On 64bit OSs, size of pointer is 8byte. Go runtime does not care about reading while writing the pointer.

Reply:
![image](https://user-images.githubusercontent.com/10111/69603018-b2bd2f80-105c-11ea-8a51-ab1476996b50.png)

Reply:
A data race is diffrent issue then assignment of a pointer being atomic.

A data race should not happn in our use case.

Reply:
Please try this.
```go
package main

import (
    "fmt"
    "runtime"
    "time"
)

type Conn struct {
    timeLocation *time.Location
}

func main() {
    runtime.GOMAXPROCS(3)
    var conn Conn
    conn.timeLocation = time.FixedZone("JST", 9)
    go func() {
        for i := 0; i < 10000; i++ {
            conn.timeLocation = time.FixedZone("JST", 9)
        }
    }()
    go func() {
        for i := 0; i < 10000; i++ {
            fmt.Println(conn.timeLocation.String())
        }
    }()

    <-make(chan struct{})
}
```

```
==================
WARNING: DATA RACE
Read at 0x00c00014a018 by goroutine 8:
  main.main.func2()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:24 +0x4f

Previous write at 0x00c00014a018 by goroutine 7:
  main.main.func1()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:19 +0x2f5

Goroutine 8 (running) created at:
  main.main()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:22 +0x37d

Goroutine 7 (running) created at:
  main.main()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:17 +0x35b
==================
==================
WARNING: DATA RACE
Read at 0x00c00004d5c0 by goroutine 8:
  time.(*Location).String()
      C:/dev/go/src/time/zoneinfo.go:93 +0x5b
  main.main.func2()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:24 +0x60

Previous write at 0x00c00004d5c0 by goroutine 7:
  time.FixedZone()
      C:/dev/go/src/time/zoneinfo.go:99 +0x7f
  main.main.func1()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:19 +0x2e8

Goroutine 8 (running) created at:
  main.main()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:22 +0x37d

Goroutine 7 (running) created at:
  main.main()
      C:/Users/mattn/go/src/github.com/mattn/go-sandbox/race/main.go:17 +0x35b
==================
```

Reply:
Right, that is a data race. Here is another example of a data race.

```
package main

import (
	"fmt"
)

var aString string

func main() {
	chanStart := make(chan struct{})
	chanEnd := make(chan struct{}, 1)
	aString = "a"

	go func() {
		<-chanStart
		aString = "b"
		fmt.Println(aString)
		chanEnd <- struct{}{}
	}()

	go func() {
		<-chanStart
		aString = "c"
		fmt.Println(aString)
		chanEnd <- struct{}{}
	}()

	close(chanStart)
	<-chanEnd
	<-chanEnd
}
```

The output could be any of the following (newlines removed):
bb, bc, cb, cc

But it will never print an a.

So in our case it may have a data **read** race and use the old time location instead of the new time location, but over a small amount of time everything will use the newly updated time location, and that is all we need.

Also note that because of the locks outside of the Go routine, a data **write** race should never happen. The go routine should never kick off more than once every 5 minutes and the query timeout is 2 minutes.

Reply:
Go runtime does not guarantee that copying "b" or "c" to aString atomically. Go runtime does not have codes to guard data copying if two goroutine access the data. On 64bit architechture, size of pointer is 8byte. While the code is reading 8byte address, the data may be broken. For example, case that the address before copying is DEADBEEF-DEADBEEF and the address after copied is BEEFCAFE-BEEFCAFE. The code may read BEEFCASE-DEADBEEF.

Reply:
https://golang.org/ref/mem

A single machine word, 8 bytes on 64 bit machine, are atomically written. A pointer on a 64 bit machine are 8 bytes.

Reply:
Ah, I was confused. Sorry. About word size thing, you are right. Mistaken simple calculation 8*8. However I want pass tests with -race.

Reply:
Race has been fixed

Reply:
Thanks. I'll check in later on my server.